cpp/src/reductions/reduction_operators.cuh

```diff
@@ -1,58 +1,183 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CUDF_REDUCTION_OPERATORS_CUH
+#define CUDF_REDUCTION_OPERATORS_CUH
+
 #include <cudf/cudf.h>
 #include <utilities/cudf_utils.h>
 #include <utilities/wrapper_types.hpp>
 #include <utilities/error_utils.hpp>
+#include <iterator/iterator.cuh>
+
+#include <utilities/device_operators.cuh>
 
-#include <utilities/device_atomics.cuh>
+#include <cmath>
 
 namespace cudf {
 namespace reductions {
 
-// ------------------------------------------------------------------------
-// difinitions of device struct for binary operation
+template<typename T>
+struct meanvar_no_count
+{
+    T value;                /// the value
+    T value_squared;        /// the value of squared
 
-struct IdentityLoader {
-    template<typename T>
-    __device__
-        T operator() (const T *ptr, int pos) const {
-        return ptr[pos];
-    }
+    CUDA_HOST_DEVICE_CALLABLE
+    meanvar_no_count(T _value=0, T _value_squared=0)
+    : value(_value), value_squared(_value_squared)
+    {};
+
+    using this_t = cudf::reductions::meanvar_no_count<T>;
+
+    CUDA_HOST_DEVICE_CALLABLE
+    this_t operator+(this_t const &rhs) const
+    {
+        return this_t(
+            (this->value + rhs.value),
+            (this->value_squared + rhs.value_squared)
+        );
+    };
+
+    CUDA_HOST_DEVICE_CALLABLE
+    bool operator==(this_t const &rhs) const
+    {
+        return (
+            (this->value == rhs.value) &&
+            (this->value_squared == rhs.value_squared)
+        );
+    };
 };
 
-struct SquaredLoader {
-    template<typename T>
-    __device__
-    T operator() (const T* ptr, int pos) const {
-        T val = ptr[pos];   // load
-        return val * val;   // squared
-    }
+template<typename T_element>
+struct transformer_meanvar_no_count
+{
+    using T_output = cudf::reductions::meanvar_no_count<T_element>;
+
+    CUDA_HOST_DEVICE_CALLABLE
+    T_output operator() (T_element const & value)
+    {
+        return T_output(value, value*value);
+    };
 };
 
+// ------------------------------------------------------------------------
+// difinitions of device struct for binary operation
+
 struct ReductionSum {
-    using Loader = IdentityLoader;
     using Op = cudf::DeviceSum;
 };
 
 struct ReductionProduct {
-    using Loader = IdentityLoader;
     using Op = cudf::DeviceProduct;
 };
 
 struct ReductionSumOfSquares {
-    using Loader = SquaredLoader;
     using Op = cudf::DeviceSum;
 };
 
 struct ReductionMin{
-    using Loader = IdentityLoader;
     using Op = cudf::DeviceMin;
 };
 
 struct ReductionMax{
-    using Loader = IdentityLoader;
     using Op = cudf::DeviceMax;
 };
 
+// operator for `mean`
+struct ReductionMean{
```

Comment:
Since these are all in the `reductions` namespace, the `Reduction*` prefix is unnecessary.